### ISSUE TITLE
Using merge on a variable that is null results in a panic

### ISSUE THREAD

**Issue by pselle** [hashicorp/terraform#21695]
Relocating from this comment: https://github.com/hashicorp/terraform/issues/5471#issuecomment-500724481

### Terraform Version

```
...
```

### Terraform Configuration Files

```hcl
...
```

### Crash Output

```
Error: Error in function call

  on ../aws-iam-role/main.tf line 20, in locals:
  20:   iam_role_tags                       = "${merge(var.iam_role_tags, map("Name", format("%s", local.iam_role_name)), map("terraform", "true"))}"
    |----------------
    | local.iam_role_name is "rds-config-role"
    | var.iam_role_tags is null

Call to function "merge" failed: panic in function implementation: can't use
ElementIterator on null value
goroutine 325 [running]:
runtime/debug.Stack(0xc00066dfa8, 0x1856a20, 0x1f7f5d0)
```
### Expected Behavior

Omit the null values

### Actual Behavior

A panic!
### References

- https://github.com/hashicorp/terraform/issues/5471
- https://github.com/hashicorp/terraform/issues/20634

**Comment by apparentlymart** [comment]
In this case it looks like the map _itself_ is what is `null`, so I think it is expected that this would return an error, but of course it should be a user-friendly error rather than a panic like this.

Curiously, I see that [this function is explicitly opting in to receiving nulls](https://github.com/hashicorp/terraform/blob/aab831611f27cf6a691da88d4a2fa9528060a967/lang/funcs/collection.go#L879), but then failing to handle them. The easiest fix here would be to remove `AllowNull: true` and thus accept the standard behavior of having the language runtime itself detect and reject the null value.

While in principle ignoring the null values altogether could be a reasonable behavior, elsewhere in the language we've tended towards being explicit about treating null as distinct from an empty value because it tends to make it easier to catch mistakes where something is mistakenly unset. I suggest that we stick with that principle here and make this an error; in line with our usual "explicit is better than implicit" principle, there is an explicit way to indicate the behavior of treating a null as if it were an empty map:

```hcl
coalesce(var.iam_role_tags, {})
```

**Comment by mikalai-t** [comment]
Just filling the gaps
### Terraform version
```
./terraform-0.12.1 --version
Terraform v0.12.1
+ provider.null v2.1.2
```
### Terraform Configuration Files
```
tree ./ -L 1
./
├── terraform-0.12.1
└── test.tf
```
#### test.tf
```
variable "empty_map" {
  type = "map"
  default = {}
}

locals {
  modify_var_before_use  = merge(var.empty_map, map("some_key", "some_value"))
}

data "null_data_source" "values" {
  inputs = {
    test_local_var = local.modify_var_before_use["some_key"]
  }
}

output "value_from_null_data_source" {
  value = data.null_data_source.values.outputs["test_local_var"]
}
```
#### Default:
```hcl
./terraform-0.12.1 apply
data.null_data_source.values: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

value_from_null_data_source = some_value
```
#### With `null` value:
```hcl
./terraform-0.12.1 apply -var=empty_map=null

Error: Error in function call

  on test.tf line 7, in locals:
   7:   modify_var_before_use  = merge(var.empty_map, map("some_key", "some_value"))
    |----------------
    | var.empty_map is null

Call to function "merge" failed: panic in function implementation: can't use
ElementIterator on null value
goroutine 46 [running]:
runtime/debug.Stack(0xc000146000, 0x1856a20, 0x1f7f5d0)
	/opt/go/src/runtime/debug/stack.go:24 +0x9d
github.com/zclconf/go-cty/cty/function.errorForPanic(...)
...
```

**Comment by pselle** [comment]
closed by #21721
